### PR TITLE
avoid using FEDataManager in FEDataInterpolation

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: ibtk/include/ibtk/FEDataInterpolation.h [hunk 1]
```diff
@@ -235,7 +243,7 @@ class FEDataInterpolation
     size_t registerInterpolatedSystem(const libMesh::System& system,
                                       const std::vector<int>& vars = std::vector<int>(1, 0),
                                       const std::vector<int>& grad_vars = std::vector<int>(),
-                                      libMesh::NumericVector<double>* system_data = nullptr);
+                                      libMesh::NumericVector<double>* system_vec = nullptr);
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataInterpolation.h:246]: There is a class named `SystemData` so I thought I would try to make it clearer that this is the vector of DOFs.

**Comment by drwells** [comment]
Makes sense - we only use `FEDataManager` in two places in this class.

Could we get rid of the whole 'use IB ghosted' logic and FEDataManager completely and just rely on the user to provide a ghosted vector?

**Comment by boyceg** [comment]
That would require more invasive changes. For instance, we evaluate the stress tensors in the IB ghosted representation in some places, and the stress tensors themselves may have dependencies on `Systems`, so we would need to go in and manually communicate to IB ghosted form.

The current implementation is not so pretty but I think it is simpler than requiring callers to figure out which vectors to pass in.

**Comment by boyceg** [comment]
Actually I think it wouldn't be too hard to make a little converter. Let me give it a shot.

**Comment by boyceg** [comment]
OK I think I did it.

**Comment by boyceg** [comment]
Jenkins retest this please

**Comment by boyceg** [comment]
@drwells test suite seems to work for me. Can you check and let me know if there are other changes?

File: src/IB/IBFEMethod.cpp [hunk 2]
```diff
@@ -330,6 +330,27 @@ libmesh_restart_file_name(const std::string& restart_dump_dirname,
...
+    std::vector<SystemData> ghosted_system_data;
+    EquationSystems* equation_systems = fe_data_manager->getEquationSystems();
+    for (const SystemData& system_data : unghosted_system_data)
+    {
+        const std::string& system_name = system_data.system_name;
+        const System& system = equation_systems->get_system(system_name);
+        NumericVector<double>* original_system_vec = system_data.system_vec;
+        if (!original_system_vec) original_system_vec = system.current_local_solution.get();
+        NumericVector<double>* ghost_system_vec =
+            fe_data_manager->buildGhostedSolutionVector(system_name, /*synch_data*/ false);
```
Comment (drwells) [inline src/IB/IBFEMethod.cpp:346]: The logic in this function partially duplicates the logic in `LibMeshSystemVectors. Could we set up a private utility function in IBFEMethod that does the same thing here but uses that class to access vectors? That would also allow us to not use `buildGhostedSolutionVector` here, which is a deprecated function.

**Comment by boyceg** [comment]
Which tests?

**Comment by boyceg** [comment]
My original rewrite here split the class into base/derived, but if we are just going to put the onus of setting up ghosting correctly on the caller, less invasive changes are needed. I just reverted most of the changes. @drwells can you see if this at least passes the test suite? I can work on switching to use `LibMeshSystemVectors`.

**Comment by drwells** [comment]
The test failure is related to something else - I'm working on a fix now.

**Comment by drwells** [comment]
> but if we are just going to put the onus of setting up ghosting correctly on the caller, less invasive changes are needed

Now seems like a good time to reach a consensus on this issue - I think that we should *always* require the caller to provide sufficient ghost information. This creates more work for the user but makes logic a lot clearer and prevents unnecessary blocks in the multipart case.

I would like to apply this principle everywhere - what do others think?

**Comment by boyceg** [comment]
> > but if we are just going to put the onus of setting up ghosting correctly on the caller, less invasive changes are needed
> 
> Now seems like a good time to reach a consensus on this issue - I think that we should _always_ require the caller to provide sufficient ghost information. This creates more work for the user but makes logic a lot clearer and prevents unnecessary blocks in the multipart case.
> 
> I would like to apply this principle everywhere - what do others think?

Sounds good to me.

**Comment by boyceg** [comment]
I just rebased off master to grab the updated test output.

**Comment by drwells** [comment]
All the `IBFE/explicit_ex4` tests pass so this is presently working.

File: src/IB/IBFEMethod.cpp (hunk 2 above)
Comment (boyceg) [reply src/IB/IBFEMethod.cpp]: The way that I could see to use `LibMeshSystemVectors` seemed too complicated, and so I just changed this to use `buildIBGhostedVector` and to return a vector of `unique_ptr`s to the built vectors.

Comment (boyceg) [reply src/IB/IBFEMethod.cpp]: (One issue is that functions associated with different parts could have different systems associated with them, and so it is not clear that `LibMeshSystemVectors` is the best way to organize things. The `LibMeshSystemVectors` interface is seems best suited for cases where all of the systems do more-or-less the same thing.)

**Comment by boyceg** [comment]
if this looks OK I can squash

File: src/IB/IBFEMethod.cpp (hunk 2 above)
Comment (drwells) [reply src/IB/IBFEMethod.cpp]: Rats. Well, it was worth trying.

**Comment by drwells** [comment]
This looks good. I would prefer to ultimately get all of the code that uses the IB partitioning out of IBFEMethod but that is far beyond the scope of this PR.

I the test suite passes on 3dafc8c0cd9cb200e058cfdc14fc9145116f69d2. Let me know if you change anything when you squash.

**Comment by boyceg** [comment]
I merely squashed down to a single commit.

**Pull request merged**